example/tpm_ibmtss2/README.md

```diff
@@ -0,0 +1,93 @@
+Test script for TPM 2
+
+Usage
+-----
+
+The TPM 2 test script loads the RSA key into the TPM's NULL hierarchy. Please
+make sure that the NULL hierarchy has no password set and no other
+application is using the TPM 2 and sufficient space is available in the TPM for
+loading the RSA key. Since access to TPM devices is typically restricted to the
+root user, the test script (time_decrypt) will need to be run as root.
+
+The TPM 2 test script can for example be run against swtpm using the vtpm
+proxy device:
+
+ > modprobe tpm_vtpm_proxy
+ > swtpm chardev --vtpm-proxy --tpmstate dir=./ --tpm2
+ New TPM device: /dev/tpm1 (major/minor = 253/1)
+
+For the test script to use the proper device, which would be /dev/tpmrm1,
+set the following environment variable:
+
+ TPM_DEVICE=/dev/tpmrm1
+
+
+When using a hardware TPM 2 the following should be used:
+
+ TPM_DEVICE=/dev/tpmrm0
+
+The following packages are needed:
+
+ - openssl-devel / libssl-dev
+ - tss2-devel / libtss2-dev
+ - swtpm-tools
+
+Run `step0.sh`, `step1.sh` as normal. Instead of running `step2.sh` run
+the `step2-marvin.sh` script.
+
+Compile this reproducer:
+```
+gcc -o time_decrypt time_decrypt.c -lcrypto -libmtss
+```
+
+Execute it against one of the `pms_values.bin` files, for either 1024
+(-n 128) or 2048 bit (-n 256):
+```
+TPM_DEVICE=/dev/tpmrm1 ./time_decrypt -i rsa2048_repeat/pms_values.bin \
+-o rsa2048_repeat/raw_times.bin -k rsa2048/pkcs8.pem -n 256
+```
+
+Note:
+If swtpm is used with a version of OpenSSL that does not implement implicit
+rejection ('older' versions), then the TPM 2 will return error messages when
+a decryption  failure occurs (e.g., bad padding) and the output may show the
+following:
+
+```
+finished  total: 22000 decryption failures: 16000  unexpected msg len: 0
+```
+
+Further, newer versions of OpenSSL will not indicate the decryption failure but
+return a message anyway, though its length will be unexpected or may have bad
+content (with expected length):
+
+```
+finished  total: 22000 decryption failures: 0  unexpected msg len: 15938
```

Comment:
I think it would be better to be explicit what kind of output is expected when.

That if you get this kind of output, that means you're running _wrong_ test vectors for the test, as that indicates that the implicit rejection is working but the test vectors are used for regular RSA decryption.

under regular (legacy) decryption both decryption failures and unexpected lengths are expected, while with implicit rejection with correct test vectors no unexpected lengths are expected

Reply:
The description reflects these results here:

```
libtpms: v0.10 + openssl 3.1.1:
finished  total: 22000 decryption failures: 0  unexpected msg len: 0

libtpms: v0.9.7 + openssl 3.1.1:
finished  total: 22000 decryption failures: 0  unexpected msg len: 0

libtpms + openssl 1.1.1l:
finished  total: 22000 decryption failures: 16000  unexpected msg len: 0
```

Reply:
No, when testing code without implicit rejection, it's expected that the decryption will return sizes other than 48 bytes:
https://github.com/tomato42/marvin-toolkit/blob/a970ff183f4191926c67dbe4d50ce9b96f504a3e/step2-alt.sh#L21

when testing without implicit rejection and with non-Marvin vectors, the expected result is non-zero decryption failures, and non-zero unexpected msg len

when testing without implicit rejection and with Marvin vectors (incorrect), the result will be non-zero decryption failures and zero unexpected msg len

when testing with implicit rejection and with non-Marvin vectors (incorrect), the result will be zero decryption failures and non-zero unexpected msg len

when testing with implicit rejection and with Marvin vectors, the expected result is zero decryption failures and zero unexpected msg lengths

Reply:
Pushed an update